drivers/modem/hl7800.c

```diff
@@ -0,0 +1,4830 @@
+		   ictx.allowSleep) {
+		PRINT_NOT_AWAKE_MSG;
+		/* If the device is sleeping (not ready to receive commands)
+		 *  then the device may send +KSUP when waking up.
+		 *  We should wait for it.
+		 */
+		ictx.wait_for_KSUP = true;
+		ictx.wait_for_KSUP_tries = 0;
+
+		set_sleep_state(HL7800_SLEEP_STATE_ASLEEP);
+		k_sem_reset(&ictx.mdm_awake);
+	}
+#endif
+}
+
+static int hl7800_RX_lock(void)
+{
+	Z_LOG(HL7800_RX_LOCK_LOG_LEVEL, "Locking RX [%p]...", k_current_get());
+	int rc = k_sem_take(&hl7800_RX_lock_sem, K_FOREVER);
+
+	if (rc != 0) {
+		LOG_ERR("Unable to lock hl7800 (%d)", rc);
+	} else {
+		Z_LOG(HL7800_RX_LOCK_LOG_LEVEL, "Locked RX [%p]",
+		      k_current_get());
+	}
+
+	return rc;
+}
+
+static void hl7800_RX_unlock(void)
+{
+	Z_LOG(HL7800_RX_LOCK_LOG_LEVEL, "UNLocking RX [%p]...",
+	      k_current_get());
+	k_sem_give(&hl7800_RX_lock_sem);
+	Z_LOG(HL7800_RX_LOCK_LOG_LEVEL, "UNLocked RX [%p]", k_current_get());
+}
+
+static bool hl7800_RX_locked(void)
+{
+	if (k_sem_count_get(&hl7800_RX_lock_sem) == 0) {
+		return true;
+	} else {
+		return false;
+	}
+}
+
+static int hl7800_TX_lock(void)
+{
+	Z_LOG(HL7800_TX_LOCK_LOG_LEVEL, "Locking TX [%p]...", k_current_get());
+	int rc = k_sem_take(&hl7800_TX_lock_sem, K_FOREVER);
+
+	if (rc != 0) {
+		LOG_ERR("Unable to lock hl7800 (%d)", rc);
+	} else {
+		Z_LOG(HL7800_TX_LOCK_LOG_LEVEL, "Locked TX [%p]",
+		      k_current_get());
+	}
+
+	return rc;
+}
+
+static void hl7800_TX_unlock(void)
+{
+	Z_LOG(HL7800_TX_LOCK_LOG_LEVEL, "UNLocking TX [%p]...",
+	      k_current_get());
+	k_sem_give(&hl7800_TX_lock_sem);
+	Z_LOG(HL7800_TX_LOCK_LOG_LEVEL, "UNLocked TX [%p]", k_current_get());
+}
+
+static bool hl7800_TX_locked(void)
+{
+	if (k_sem_count_get(&hl7800_TX_lock_sem) == 0) {
+		return true;
+	} else {
+		return false;
+	}
+}
+
+static void hl7800_lock(void)
+{
+	hl7800_TX_lock();
+	hl7800_RX_lock();
+}
+
+static void hl7800_unlock(void)
+{
+	hl7800_RX_unlock();
+	hl7800_TX_unlock();
+}
+
+static struct hl7800_socket *socket_get(void)
+{
+	int i;
+	struct hl7800_socket *sock = NULL;
+
+	for (i = 0; i < MDM_MAX_SOCKETS; i++) {
+		if (!ictx.sockets[i].context) {
+			sock = &ictx.sockets[i];
+			break;
+		}
+	}
+
+	return sock;
+}
+
+static struct hl7800_socket *socket_from_id(int socket_id)
+{
+	int i;
+	struct hl7800_socket *sock = NULL;
+
+	if (socket_id < 1) {
+		return NULL;
+	}
+
+	for (i = 0; i < MDM_MAX_SOCKETS; i++) {
+		if (ictx.sockets[i].socket_id == socket_id) {
+			sock = &ictx.sockets[i];
+			break;
+		}
+	}
+
+	return sock;
+}
+
+static void socket_put(struct hl7800_socket *sock)
+{
+	if (!sock) {
+		return;
+	}
+
+	sock->context = NULL;
+	sock->socket_id = -1;
+	sock->created = false;
+	sock->reconfig = false;
+	sock->error = false;
+	sock->error_val = -1;
+	sock->rx_size = 0;
+	sock->state = SOCK_IDLE;
+	(void)memset(&sock->src, 0, sizeof(struct sockaddr));
+	(void)memset(&sock->dst, 0, sizeof(struct sockaddr));
+}
+
+char *hl7800_sprint_ip_addr(const struct sockaddr *addr)
```

Comment:
> Tried using `net_sprint_addr()` but it prints the address backwards and causes issues.

The `net_sprint_addr()` expects that the IP address values are in network byte order. If the IP address is printed in reverse order, then it means that the sockaddr values are in host byte order, which would be very surprising.
Both this helper and `net_sprint_addr()` use `net_addr_ntop()` internally, so it sounds a bit weird if they would print the addresses differently.

Reply:
I found the difference.  `net_addr_ntop` expects a `struct in_addr *`. My function `hl7800_sprint_ip_addr` expects `struct sockaddr *`.  Leaving my function as a wrapper because it is more convinient to use because it determines the correct `in_addr` or `in6_addr` to use for me.